### DISCUSSION TITLE
64-bit Windows, Oracle 19c, Warning: a running Virtual User was terminated, any pending output has been discarded.

### DISCUSSION THREAD

**Discussion by UniqueN4me** [TPC-Council/HammerDB#223] [Q&A] [answered]
I've followed the setup/install directions you posted including configuring the sqlnet.ora file and adding an ORACLE_LIBRARY system variable followed by a reboot.  I'm having the exact same problem as this guy:

https://sourceforge.net/p/hammerdb/discussion/292312/thread/5b93ad5d/

Where, on schema build of a TPROC-C benchmark, the system immediately errors out.  hammerdbcli library check shows the oracletcl library as good and when I run the instructions in the thread, I get the same thing that guy did:

(HammerDB-4.0) 11 % package require Oratcl
4.6
(HammerDB-4.0) 12 % set connect SYSTEM/test123@myserver
SYSTEM/test123@myserver
(HammerDB-4.0) 13 % set lda [oralogon $connect]

(HammerDB-4.0) 14 % puts $lda
can't read "lda": no such variable
(HammerDB-4.0) 15 % 

I'm on 64-bit Windows 10 with 64-bit HammerDB and 64-bit instantclient, so there's no issue with mixing those up.  I love your tool and I've used it a few times before, but I always struggle with the initial setup...

ORACLE_LIBRARY is set to C:\oracle\instantclient_19_10, which contains the oci.dll as well as a sqlnet.ora file with the required settings and a tnsnames.ora file with the connection information.

**Comment by sm-shaw** [comment]
Hi,

In some cases (open source) we could compile the client libraries statically (that might even be permitted with the Oracle instant client?) however that gives us less flexibility to upgrade the database client and therefore loading dynamically requires more setup but gives us this flexibility especially if there are bugs in the client.  Loading dynamically means that the library needs to be installed and HammerDB needs to know where to find it. 

Firstly once the library is installed the PATH environment variable should be set to the location of oci.dll so for example on my system this is in 
`E:\app\client\Steve\product\18.0.0\client_1\bin`
and my path environment variable includes this directory
![oracleclient](https://user-images.githubusercontent.com/38044085/113303373-ed8e4200-92f8-11eb-8864-2c84c8367d8b.PNG)
In 
`E:\app\client\Steve\product\18.0.0\client_1\network\admin`
I have a sqlnet.ora including the settings for the long standing Oracle bug
```
For the instant client the ORACLE_LIBRARY environment variable can also be set to the same location.

# sqlnet.ora Network Configuration File: E:\app\client\Steve\product\18.0.0\client_1\NETWORK\ADMIN\sqlnet.ora
# Generated by Oracle configuration tools.

# This file is actually generated by netca. But if customers choose to 
# install "Software Only", this file wont exist and without the native 
# authentication, they will not be able to connect to the database on NT.

SQLNET.AUTHENTICATION_SERVICES= (NTS)

NAMES.DIRECTORY_PATH= (TNSNAMES, EZCONNECT)
SQLNET.AUTHENTICATION_SERVICES = (NTS)
DIAG_ADR_ENABLED=OFF DIAG_SIGHANDLER_ENABLED=FALSE
DIAG_DDE_ENABLED=FALSE
```
and an example tnsnames.ora for container and pluggable

```
VULCDB1 =
  (DESCRIPTION =
  (ADDRESS_LIST =
    (ADDRESS = (PROTOCOL = TCP)(HOST = 192.168.1.213)(PORT = 1521)))
    (CONNECT_DATA =
      (SERVER = DEDICATED)
      (SERVICE_NAME = VULCDB1)
    )
  )

VULPDB1 =
  (DESCRIPTION =
  (ADDRESS_LIST =
    (ADDRESS = (PROTOCOL = TCP)(HOST = 192.168.1.213)(PORT = 1521)))
    (CONNECT_DATA =
      (SERVER = DEDICATED)
      (SERVICE_NAME = vulpdb1)
    )
  )

LISTENER =
  (ADDRESS = (PROTOCOL = TCP)(HOST = vulture)(PORT = 1521))
```
Opening a regular command window you should be able to do tnsping to your database (mine is running on the same host under the excellent Virtual Box https://www.virtualbox.org/ ). If it can't make sure the listener is configured and running. 

```
C:\Users\Steve>tnsping vulcdb1

TNS Ping Utility for 64-bit Windows: Version 18.0.0.0.0 - Production on 01-APR-2021 15:00:39

Copyright (c) 1997, 2018, Oracle.  All rights reserved.

Used parameter files:
E:\app\client\Steve\product\18.0.0\client_1\network\admin\sqlnet.ora

Used TNSNAMES adapter to resolve the alias
Attempting to contact (DESCRIPTION = (ADDRESS_LIST = (ADDRESS = (PROTOCOL = TCP)(HOST = 192.168.1.213)(PORT = 1521))) (CONNECT_DATA = (SERVER = DEDICATED) (SERVICE_NAME = VULCDB1)))
OK (0 msec)
```
Then open the HammerDB CLI and run librarycheck, the example below shows it can be found.
```
HammerDB CLI v4.0
Copyright (C) 2003-2020 Steve Shaw
Type "help" for a list of commands
The xml is well-formed, applying configuration
hammerdb>librarycheck
Checking database library for Oracle
Success ... loaded library Oratcl for Oracle
Checking database library for MSSQLServer
Success ... loaded library tdbc::odbc for MSSQLServer
Checking database library for Db2
Success ... loaded library db2tcl for Db2
Checking database library for MySQL
Success ... loaded library mysqltcl for MySQL
Checking database library for PostgreSQL
Success ... loaded library Pgtcl for PostgreSQL

hammerdb>
```
We can now use the CLI to test manually,  The example below shows we successfully loaded the Oratcl package and can login to the database.
```
HammerDB CLI v4.0
Copyright (C) 2003-2020 Steve Shaw
Type "help" for a list of commands
The xml is well-formed, applying configuration
hammerdb>package require Oratcl
4.6
hammerdb>set connect system/oracle@VULPDB1
system/oracle@VULPDB1
hammerdb>set lda [ oralogon $connect ]
oratcl0
hammerdb>
```
The important aspect is making sure that HammerDB has the exact location of the oci.dll file.

**Reply by UniqueN4me** [reply]
I feel like I did all this, but I'll do a reinstall and run through it all again fresh.  Thanks for putting all this together!

**Comment by sm-shaw** [comment]
Moved Issue to discussion as a configuration/setup issue rather than a HammerDB issue that needs a pull request to resolve.

**Comment by UniqueN4me** [comment, marked as answer]
tnsping was originally working for me, but I have 4 versions of the instantclient on my machine, so instead of calling it from $path, I called it directly from the file directory it resided in and I found that tnsping was missing an $ORACLE_HOME variable.  I added $ORACLE_HOME pointing to the corresponding home directory and restarted the machine, after which the tnsping that hammerdb was using worked correctly and that resolved my issue.

It helped following your directions very closely, thank you!

**Reply by sm-shaw** [reply]
That is great many thanks for the follow up, that adds so much value to an issue/discussion topic to say that the answer resolved the problem and then helps other people experiencing the same thing in future.
